### PR TITLE
Issue #9857: Add first test cases for FillRequestHandler.

### PR DESCRIPTION
**Pull request by pocmo**
Okay, here are the first tests for our Android Autofill implementation. Unfortunately the Autofill API classes can't be constructed or inspected. So I had to create this intermediate representation that we translate between. The Lockwise team already built this great XML representation that we can use in tests and I took a bunch of that code.

I haven't written tests for all XML files from the Lockwise repository yet. In addition to that I would like to collect some more samples from apps. However this is already good to go and there's no need to wait for more tests.

### CODE REVIEW

File: components/feature/autofill/src/main/java/mozilla/components/feature/autofill/AbstractAutofillService.kt [hunk 1]
```diff
@@ -35,7 +36,14 @@ abstract class AbstractAutofillService : AutofillService() {
         // seems to get destroyed before we invoke a method on the callback. So we need a scope that
         // lives longer than the service.
         GlobalScope.launch(Dispatchers.IO) {
-            val response = fillHandler.handle(request)
+            // You may be wondering why we translate the AssistStructure into a RawStructure and then
+            // create a FillResponseBuilder that outputs the FillResponse. This is purely for testing.
+            // Neither AssistStructure nor FillResponse can be created by us and they do not let us
+            // inspect their data. So we create this intermediate objects that we can create and
```
Comment (grigoryk) [inline components/feature/autofill/src/main/java/mozilla/components/feature/autofill/AbstractAutofillService.kt:42]: nit: s/this/these

File: components/feature/autofill/src/main/java/mozilla/components/feature/autofill/handler/FillRequestHandler.kt [hunk 2]
```diff
@@ -88,7 +72,7 @@ internal class FillRequestHandler(
      * [AssistStructure] is `null`
      */
     @Suppress("ReturnCount")
-    suspend fun handleConfirmation(structure: AssistStructure?, loginId: String): Dataset? {
+    suspend fun handleConfirmation(structure: RawStructure?, loginId: String): DatasetBuilder? {
```
Comment (grigoryk) [inline components/feature/autofill/src/main/java/mozilla/components/feature/autofill/handler/FillRequestHandler.kt:75]: Update the kdoc w/ the new return type.

Comment (grigoryk) [inline components/feature/autofill/src/main/java/mozilla/components/feature/autofill/handler/FillRequestHandler.kt:75]: I think making RawStructure param as non-nullable would be a bit cleaner. Not much you can really do w/o a structure, so why allow it as `null`?

File: components/feature/autofill/src/main/java/mozilla/components/feature/autofill/ui/AbstractAutofillConfirmActivity.kt [hunk 3]
```diff
@@ -49,7 +50,8 @@ abstract class AbstractAutofillConfirmActivity : FragmentActivity() {
 
         // While the user is asked to confirm, we already try to build the fill response asynchronously.
         dataset = lifecycleScope.async(Dispatchers.IO) {
-            fillHandler.handleConfirmation(structure, loginId)
+            val builder = fillHandler.handleConfirmation(structure?.toRawStructure(), loginId)
```
Comment (grigoryk) [inline components/feature/autofill/src/main/java/mozilla/components/feature/autofill/ui/AbstractAutofillConfirmActivity.kt:53]: If the `structure` is somehow `null` here, we can avoid invoking `lifecycleScope.async` entirely, right?

File: components/feature/autofill/src/main/java/mozilla/components/feature/autofill/ui/AbstractAutofillUnlockActivity.kt [hunk 4]
```diff
@@ -40,7 +41,10 @@ abstract class AbstractAutofillUnlockActivity : FragmentActivity() {
         val structure: AssistStructure? = intent.getParcelableExtra(AutofillManager.EXTRA_ASSIST_STRUCTURE)
 
         // While the user is asked to authenticate, we already try to build the fill response asynchronously.
-        fillResponse = lifecycleScope.async(Dispatchers.IO) { fillHandler.handle(structure, forceUnlock = true) }
+        fillResponse = lifecycleScope.async(Dispatchers.IO) {
+            val builder = fillHandler.handle(structure?.toRawStructure(), forceUnlock = true)
```
Comment (grigoryk) [inline components/feature/autofill/src/main/java/mozilla/components/feature/autofill/ui/AbstractAutofillUnlockActivity.kt:45]: Same comment about `null` structure.

Also, do you care about asserting that we actually get a structure? Not worth crashing over, but could be useful to perhaps log this as a `w`, at least, to maybe help w/ debugging weird issues.

File: components/feature/autofill/src/test/java/mozilla/components/feature/autofill/test/MockStructure.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import mozilla.components.feature.autofill.structure.AutofillNodeNavigator
+import mozilla.components.feature.autofill.structure.RawStructure
+import java.io.File
+
+internal fun FillRequestHandlerTest.createMockStructure(filename: String, packageName: String): RawStructure {
+    val classLoader = javaClass.classLoader ?: throw RuntimeException("No class loader")
+    val resource = classLoader.getResource(filename) ?: throw RuntimeException("Resource not found")
+    val file = File(resource.path)
+
+    return object : RawStructure {
```
Comment (grigoryk) [inline components/feature/autofill/src/test/java/mozilla/components/feature/autofill/test/MockStructure.kt:18]: Why an anonymous object, and not, say, a `TestRawStructure`? Similarly, `AssistStructure.toRawStructure` could return something like `AndroidRawStructure` or something.

That could make stacktraces a little bit more friendly?

**Review by grigoryk — APPROVED** [review]
Just some minor comments, looks fine overall!

File: components/feature/autofill/src/main/java/mozilla/components/feature/autofill/handler/FillRequestHandler.kt (hunk 2 above)
Comment (pocmo) [reply components/feature/autofill/src/main/java/mozilla/components/feature/autofill/handler/FillRequestHandler.kt:75]: The reason was for that check to be testable. If I make this non-nullable then this `null` check needs to go into `AbstractAutofillService` and there it cannot be tested.

**Pull request merged**